// @ironsoftware/ironpdf/src/internal/generated_proto/iron_pdf_service.ts
import type * as grpc from '@grpc/grpc-js';
import type { MessageTypeDefinition } from '@grpc/proto-loader';

import type { IronPdfServiceClient as _ironpdfengineproto_IronPdfServiceClient, IronPdfServiceDefinition as _ironpdfengineproto_IronPdfServiceDefinition } from './ironpdfengineproto/IronPdfService';

type SubtypeConstructor<Constructor extends new (...args: any) => any, Subtype> = {
  new(...args: ConstructorParameters<Constructor>): Subtype;
};

export interface ProtoGrpcType {
  google: {
    protobuf: {
      Timestamp: MessageTypeDefinition
    }
  }
  ironpdfengineproto: {
    BooleanResultP: MessageTypeDefinition
    BytesResultStreamP: MessageTypeDefinition
    ChromeAddHtmlHeaderFooterRequestP: MessageTypeDefinition
    ChromeApplyCookiesRequestP: MessageTypeDefinition
    ChromeApplyStampRequestStreamP: MessageTypeDefinition
    ChromeBarcodeEncodingP: MessageTypeDefinition
    ChromeBarcodeStamperInfoP: MessageTypeDefinition
    ChromeCefBrowserSettingsP: MessageTypeDefinition
    ChromeCefRenderSettingsP: MessageTypeDefinition
    ChromeClearCookiesRequestP: MessageTypeDefinition
    ChromeFitToPaperModesP: MessageTypeDefinition
    ChromeHtmlHeaderFooterP: MessageTypeDefinition
    ChromeHtmlStamperInfoP: MessageTypeDefinition
    ChromeHttpLoginCredentialsP: MessageTypeDefinition
    ChromeImageBehaviorP: MessageTypeDefinition
    ChromeImageFilesToPdfRequestStreamP: MessageTypeDefinition
    ChromeImageStamperInfoP: MessageTypeDefinition
    ChromeImageToPdfRequestStreamP: MessageTypeDefinition
    ChromePdfCssMediaTypeP: MessageTypeDefinition
    ChromePdfPaperOrientationP: MessageTypeDefinition
    ChromePdfPaperSizeP: MessageTypeDefinition
    ChromePdfRenderOptionsP: MessageTypeDefinition
    ChromeRenderMeasureHtmlRequestStreamP: MessageTypeDefinition
    ChromeRenderPdfDocumentFromHtmlRequestStreamP: MessageTypeDefinition
    ChromeRenderPdfDocumentFromRtfStringRequestStreamP: MessageTypeDefinition
    ChromeRenderPdfDocumentFromUriRequestP: MessageTypeDefinition
    ChromeRenderPdfDocumentFromZipFileRequestStreamP: MessageTypeDefinition
    ChromeTableOfContentsTypesP: MessageTypeDefinition
    ChromeTextHeaderFooterP: MessageTypeDefinition
    ChromeTextStamperInfoP: MessageTypeDefinition
    ChromeUseMarginsP: MessageTypeDefinition
    EmptyP: MessageTypeDefinition
    EmptyResultP: MessageTypeDefinition
    FontTypeP: MessageTypeDefinition
    HandshakeRequestP: MessageTypeDefinition
    HandshakeResponseP: MessageTypeDefinition
    HorizontalAlignmentP: MessageTypeDefinition
    ImageResultStreamP: MessageTypeDefinition
    ImagesResultStreamP: MessageTypeDefinition
    IntListResultP: MessageTypeDefinition
    IntResultP: MessageTypeDefinition
    IronPdfService: SubtypeConstructor<typeof grpc.Client, _ironpdfengineproto_IronPdfServiceClient> & { service: _ironpdfengineproto_IronPdfServiceDefinition }
    LengthP: MessageTypeDefinition
    ListIntP: MessageTypeDefinition
    ListStringP: MessageTypeDefinition
    ListStringResultP: MessageTypeDefinition
    MeasurementUnitP: MessageTypeDefinition
    PdfBookmarkCollectionP: MessageTypeDefinition
    PdfDocumentP: MessageTypeDefinition
    PdfDocumentResultP: MessageTypeDefinition
    PdfiumAddBackgroundForegroundRequestP: MessageTypeDefinition
    PdfiumAddFormFieldAnnotationRequestP: MessageTypeDefinition
    PdfiumAddFormFieldRequestP: MessageTypeDefinition
    PdfiumAddLinkAnnotationRequestP: MessageTypeDefinition
    PdfiumAddPdfAttachmentRequestP: MessageTypeDefinition
    PdfiumAddPdfAttachmentResultP: MessageTypeDefinition
    PdfiumAddSignatureImageRequestStreamP: MessageTypeDefinition
    PdfiumAddTextAnnotationRequestP: MessageTypeDefinition
    PdfiumAddTextHeaderFooterRequestP: MessageTypeDefinition
    PdfiumAnnotationIconP: MessageTypeDefinition
    PdfiumBookmarkDescriptorCollectionP: MessageTypeDefinition
    PdfiumBookmarkDescriptorP: MessageTypeDefinition
    PdfiumBookmarkP: MessageTypeDefinition
    PdfiumCheckBoxFieldP: MessageTypeDefinition
    PdfiumComboBoxFieldP: MessageTypeDefinition
    PdfiumComboboxOptionsP: MessageTypeDefinition
    PdfiumCompressImagesRequestP: MessageTypeDefinition
    PdfiumConvertToPdfARequestStreamP: MessageTypeDefinition
    PdfiumConvertToPdfUARequestP: MessageTypeDefinition
    PdfiumCopyPagesRequestP: MessageTypeDefinition
    PdfiumDrawBitmapRequestStreamP: MessageTypeDefinition
    PdfiumDrawTextRequestP: MessageTypeDefinition
    PdfiumEditTextAnnotationRequestP: MessageTypeDefinition
    PdfiumExtractAllImagesRequestP: MessageTypeDefinition
    PdfiumExtractAllRawImagesRequestP: MessageTypeDefinition
    PdfiumExtractAllTextRequestP: MessageTypeDefinition
    PdfiumFlattenFormRequestP: MessageTypeDefinition
    PdfiumFontInfoCollectionP: MessageTypeDefinition
    PdfiumFontInfoP: MessageTypeDefinition
    PdfiumFormFieldCollectionP: MessageTypeDefinition
    PdfiumFormFieldP: MessageTypeDefinition
    PdfiumFormFieldTypesP: MessageTypeDefinition
    PdfiumGetAnnotationCountRequestP: MessageTypeDefinition
    PdfiumGetAnnotationRequestP: MessageTypeDefinition
    PdfiumGetBinaryDataRequestStreamP: MessageTypeDefinition
    PdfiumGetBookmarksDescriptorRequestP: MessageTypeDefinition
    PdfiumGetBookmarksDescriptorResultP: MessageTypeDefinition
    PdfiumGetComboboxOptionsRequestP: MessageTypeDefinition
    PdfiumGetComboboxOptionsResultP: MessageTypeDefinition
    PdfiumGetComboboxSelectedIndexRequestP: MessageTypeDefinition
    PdfiumGetComboboxSelectedIndexResultP: MessageTypeDefinition
    PdfiumGetFontInfoCollectionRequestP: MessageTypeDefinition
    PdfiumGetFontInfoCollectionResultP: MessageTypeDefinition
    PdfiumGetFormFieldIsReadOnlyRequestP: MessageTypeDefinition
    PdfiumGetFormFieldIsReadOnlyResultP: MessageTypeDefinition
    PdfiumGetFormFieldValueRequestP: MessageTypeDefinition
    PdfiumGetFormFieldValueResultP: MessageTypeDefinition
    PdfiumGetFormFieldsRequestP: MessageTypeDefinition
    PdfiumGetFormFieldsResultP: MessageTypeDefinition
    PdfiumGetFormRequestP: MessageTypeDefinition
    PdfiumGetFormResultP: MessageTypeDefinition
    PdfiumGetMetadataDictRequestP: MessageTypeDefinition
    PdfiumGetMetadataKeysRequestP: MessageTypeDefinition
    PdfiumGetMetadataRequestP: MessageTypeDefinition
    PdfiumGetPageBoundsRequestP: MessageTypeDefinition
    PdfiumGetPageCountRequestP: MessageTypeDefinition
    PdfiumGetPageCountResultP: MessageTypeDefinition
    PdfiumGetPageRequestP: MessageTypeDefinition
    PdfiumGetPageResultP: MessageTypeDefinition
    PdfiumGetPagesRequestP: MessageTypeDefinition
    PdfiumGetPagesResultP: MessageTypeDefinition
    PdfiumGetPaperSizePresetRequestP: MessageTypeDefinition
    PdfiumGetPaperSizePresetResultP: MessageTypeDefinition
    PdfiumGetPdfAttachmentCountRequestP: MessageTypeDefinition
    PdfiumGetPdfAttachmentCountResultP: MessageTypeDefinition
    PdfiumGetPdfAttachmentDataRequestP: MessageTypeDefinition
    PdfiumGetPdfAttachmentDataResultStreamP: MessageTypeDefinition
    PdfiumGetPdfAttachmentNameRequestP: MessageTypeDefinition
    PdfiumGetPdfAttachmentNameResultP: MessageTypeDefinition
    PdfiumGetPdfSecuritySettingsRequestP: MessageTypeDefinition
    PdfiumGetPdfSecuritySettingsResultP: MessageTypeDefinition
    PdfiumGetRevisionRequestP: MessageTypeDefinition
    PdfiumGetSignatureByteRangeRequestP: MessageTypeDefinition
    PdfiumGetSignatureCountRequestP: MessageTypeDefinition
    PdfiumGetSignatureCountResultP: MessageTypeDefinition
    PdfiumGetTextAnnotationResultP: MessageTypeDefinition
    PdfiumGetVerifiedSignatureRequestStreamP: MessageTypeDefinition
    PdfiumGetVerifySignatureResultP: MessageTypeDefinition
    PdfiumIDocumentObjectP: MessageTypeDefinition
    PdfiumIFormFieldAnnotationObjectP: MessageTypeDefinition
    PdfiumIFormFieldAnnotationP: MessageTypeDefinition
    PdfiumIFormFieldObjectP: MessageTypeDefinition
    PdfiumIFormFieldP: MessageTypeDefinition
    PdfiumInsertBookmarkRequestP: MessageTypeDefinition
    PdfiumLayerModesP: MessageTypeDefinition
    PdfiumMergePageObjectsAppendRequestP: MessageTypeDefinition
    PdfiumMetadataDictP: MessageTypeDefinition
    PdfiumMetadataDictResultP: MessageTypeDefinition
    PdfiumMetadataFieldP: MessageTypeDefinition
    PdfiumMetadataFieldResultP: MessageTypeDefinition
    PdfiumPageBoundsP: MessageTypeDefinition
    PdfiumPageBoundsPResultP: MessageTypeDefinition
    PdfiumPageCollectionP: MessageTypeDefinition
    PdfiumPageP: MessageTypeDefinition
    PdfiumPageRotationP: MessageTypeDefinition
    PdfiumPaperSizePresetP: MessageTypeDefinition
    PdfiumPdfAnnotationFlagsP: MessageTypeDefinition
    PdfiumPdfAttachmentCollectionP: MessageTypeDefinition
    PdfiumPdfDocumentConstructorStreamP: MessageTypeDefinition
    PdfiumPdfDocumentInsertRequestP: MessageTypeDefinition
    PdfiumPdfDocumentMergeRequestP: MessageTypeDefinition
    PdfiumPdfDocumentPermissionsP: MessageTypeDefinition
    PdfiumPdfEditSecurityP: MessageTypeDefinition
    PdfiumPdfFormFieldFlagsP: MessageTypeDefinition
    PdfiumPdfFormFieldTypeP: MessageTypeDefinition
    PdfiumPdfFormP: MessageTypeDefinition
    PdfiumPdfPrintSecurityP: MessageTypeDefinition
    PdfiumPdfSecuritySettingsP: MessageTypeDefinition
    PdfiumPdfSignatureCollectionP: MessageTypeDefinition
    PdfiumPdfSignatureP: MessageTypeDefinition
    PdfiumPdfToImagesRequestP: MessageTypeDefinition
    PdfiumPdfToMultiPageTiffImageRequestP: MessageTypeDefinition
    PdfiumPixelFormatP: MessageTypeDefinition
    PdfiumRawSignatureChunkWithIndexP: MessageTypeDefinition
    PdfiumRemoveAnnotationRequestP: MessageTypeDefinition
    PdfiumRemoveBookmarksRequestP: MessageTypeDefinition
    PdfiumRemoveFormFieldAnnotationRequestP: MessageTypeDefinition
    PdfiumRemoveFormFieldRequestP: MessageTypeDefinition
    PdfiumRemoveMetadataRequestP: MessageTypeDefinition
    PdfiumRemovePagesRequestP: MessageTypeDefinition
    PdfiumRemovePagesResultP: MessageTypeDefinition
    PdfiumRemovePasswordsAndEncryptionRequestP: MessageTypeDefinition
    PdfiumRemovePdfAttachmentRequestP: MessageTypeDefinition
    PdfiumRemoveSignaturesRequestP: MessageTypeDefinition
    PdfiumRemoveStructTreeRequestP: MessageTypeDefinition
    PdfiumRenameFieldRequestP: MessageTypeDefinition
    PdfiumRenameFormFieldResultP: MessageTypeDefinition
    PdfiumReplaceTextRequestP: MessageTypeDefinition
    PdfiumResizePageRequestP: MessageTypeDefinition
    PdfiumSetFormFieldFontRequestP: MessageTypeDefinition
    PdfiumSetFormFieldIsReadOnlyRequestP: MessageTypeDefinition
    PdfiumSetFormFieldValue2RequestP: MessageTypeDefinition
    PdfiumSetFormFieldValueRequestP: MessageTypeDefinition
    PdfiumSetMetadataDictRequestP: MessageTypeDefinition
    PdfiumSetMetadataRequestP: MessageTypeDefinition
    PdfiumSetPagesRotationRequestP: MessageTypeDefinition
    PdfiumSetPdfAttachmentDataRequestStreamP: MessageTypeDefinition
    PdfiumSetPdfSecuritySettingsRequestP: MessageTypeDefinition
    PdfiumSignRequestStreamP: MessageTypeDefinition
    PdfiumSignResultP: MessageTypeDefinition
    PdfiumSignaturePermissionsP: MessageTypeDefinition
    PdfiumStampBackgroundWatermarkRequestStreamP: MessageTypeDefinition
    PdfiumStampFooterWatermarkRequestStreamP: MessageTypeDefinition
    PdfiumTextAnnotationP: MessageTypeDefinition
    PdfiumTextFieldP: MessageTypeDefinition
    PdfiumUnknownFieldP: MessageTypeDefinition
    PdfiumUpdateFormFieldAnnotationRequestP: MessageTypeDefinition
    PdfiumUpdateFormFieldRequestP: MessageTypeDefinition
    PdfiumVerifiedSignatureCollectionP: MessageTypeDefinition
    PdfiumVerifiedSignatureP: MessageTypeDefinition
    RawImageChunkWithIndexAndFileTypeP: MessageTypeDefinition
    RawImageChunkWithIndexP: MessageTypeDefinition
    RemoteExceptionP: MessageTypeDefinition
    RenderOptionWaitForP: MessageTypeDefinition
    SizeP: MessageTypeDefinition
    SizeResultP: MessageTypeDefinition
    StringDictionaryEntryP: MessageTypeDefinition
    StringDictionaryP: MessageTypeDefinition
    StringDictionaryResultP: MessageTypeDefinition
    StringResultP: MessageTypeDefinition
    SystemIsLicensedRequestP: MessageTypeDefinition
    SystemIsValidLicenseRequestP: MessageTypeDefinition
    SystemSetIsDebugRequestP: MessageTypeDefinition
    SystemSetLicenseKeyRequestP: MessageTypeDefinition
    VerticalAlignmentP: MessageTypeDefinition
  }
}
